secret/scanner.go

```diff
@@ -11,24 +11,53 @@ import (
 var lineSep = []byte{'\n'}
 
 type Scanner struct {
-	Rules []Rule
+	Rules         []Rule
+	AllowList     AllowList
+	ExcludeBlocks ExcludeBlocks
 }
-
 type Rule struct {
-	ID       string
-	Type     types.SecretRuleType
-	Severity string
-	Title    string
-	Regex    *regexp.Regexp
-	Path     *regexp.Regexp
+	ID            string
+	Type          types.SecretRuleType
+	Severity      string
+	Title         string
+	Regex         *regexp.Regexp
+	Path          *regexp.Regexp
+	AllowList     AllowList
+	ExcludeBlocks ExcludeBlocks
+}
+
+type RuleResult struct {
+	RuleID        string
+	Type          types.SecretRuleType
+	Severity      string
+	Title         string
+	StartPosition int
+	EndPosition   int
+	Match         string
+}
+
+type AllowList struct {
+	Title   string
+	Regexes []*regexp.Regexp
+	Paths   []*regexp.Regexp
 }
 
-func NewScanner(rulePath string) Scanner {
+type ExcludeBlocks struct {
+	Title   string
+	Regexes []*regexp.Regexp
+}
+
+func NewScanner(rulePath string, rules []Rule, allowlist AllowList, excludeBlocks ExcludeBlocks) Scanner { // TODO: remove rules, allowlist, excludeBlocks when rulePath implemented
 	if rulePath != "" {
 		// TODO: Load custom rules here
 	}
+	if len(rules) == 0 {
+		rules = builtinRules
```

Comment:
Should we disable built-in rules if a user passes even a single rule? I thought we wanted to merge user-specified rules into build-in rules.

Reply:
It's only for the tests right now, I'll add reading from config YAML file and implement the merge in a different branch. 
I think we should merge the rules by the ruleID, so the user will be able to replace built-in rules if needed and add custom rules in a single config file.

Reply:
It sounds a good idea!